_sources/configuration.rst.txt
Configuration
=============
The configuration for MAH is handled via the standard Python `ConfigParser <http://docs.python.org/library/configparser.html>`_ framework, which uses the Microsoft INI file structure.


location
--------
The location of the configuration file can be specified via the environment
variable MAHCONFIG.

Example (for Bourne compatible shells):
::

    export MAHCONFIG=/path/to/config/mah.conf

sample configuration
--------------------
The following shows an example configuration file:

Example mah.conf
::

    [database]
    connect = mysql://<username>:<password>@<host>/<database>
    logsql = False

    [ldap]
    size_limit = 50 ;max count of findings
    paged_size = 10
    time_limit = 50
    server = ldap://<hostname>:<port> ;for ssl use ldaps:
    user = admin
    password = 12345
    scope = ou=users, dc=corp, dc=com
    attributes = uid,cn,title,mail,telephoneNumber,location
    filter = uid,cn,telephoneNumber
    attributes = uid,cn,title,mail,telephoneNumber,location
    id_attribute = uid
    web_attributes = Username,Full Name,Title,Email,Phone,Location


    [login]
    check_password = True
    use_radius = True
    radius_dictionary = /usr/share/doc/python-pyrad-1.1/example/dictionary
    radius_server = <hostname>
    radius_secret = <password>
    radius_nas_identifer = 1
    radius_nas_ip_address = <ip_address>

    [authentication]
    timeout = 600
    length = 9
    variable_length = True

    [report]
    email_from = mah@corp.com
    email_to = user@corp.com
    email_subject = MAH suspicious activity report
    smtp_server = smtp.corp.com

    [application]
    host = 0.0.0.0
    port = 5000
    SESSION_KEY = SomeSort0fRandomStringShouldGoHere?
    SESSION_COOKIE_SECURE = True
    SESSION_COOKIE_HTTPONLY = True
    PREFERRED_URL_SCHEME = https
    SESSION_TIMEOUT = 600
    REFRESH = 30

    [logging]
    file_name = ./mah.log
    file_level = DEBUG
    syslog_level = INFO
    syslog_facility =  local1
    syslog_host = localhost
    syslog_port = 514

configuration elements
----------------------
This section provides a description of all the configuration elements,
including their use and function. The configuration elements are described
in the form section.variable.

database.connection
```````````````````
The database connection string.
::

    <engine>://<db-user>:<db-password>@<db-host>/<database>

This is the argument provided to the create_engine() function of SQL Alchemy. For more
information, see `the SQL Alchemy documentation <http://docs.sqlalchemy.org/en/rel_0_7/core/engines.html>`_ for more information.

database.logsql
```````````````
Enable logging of SQL statements.
::

    logsql = False

Use this configuration option to enable or disable SQL logging for
debugging purposes.

ldap.size_limit
```````````````
Global limit - sets how many entries the server will examine.
Count of staff on a page.
::

    size_limit = 250

ldap.paged_size
```````````````
Global limit - sets maximum page size.
Microsoft Active Directory function.
::

    paged_size = 5

ldap.time_limit
```````````````
Global limit - sets maximum time limit.
::

        time_limit = 30

ldap.server
```````````
The LDAP server URL to connect to. For SSL use ldaps:
::

    ldap://<address>:<port>

ldap.user and ldap.password
```````````````````````````
Login credentials into LDAP server.
::

    user = admin
    password = 12345

ldap.scope
``````````
The LDAP scope in which to perform directory searches.
::

    scope = ou=users, dc=corp, dc=com

ldap.filter
```````````
The LDAP filter specify final results based on terms.
::

    filter =  uid,cn,telephoneNumber

ldap.attributes
```````````````
The LDAP attributes specify directory searches.
::

    attributes = cn,uid,title,mail,telephoneNumber,location

ldap.id_attribute
`````````````````
The LDAP id_attribute specify search_by_uid attribute for searches.
::

    id_attribute = uid

ldap.web_attributes
```````````````````
The LDAP web_attributes specify description of every column in table on web
page. Used in the same order as 'ldap.attributes'
::

    web_attributes = Username,Full Name,Title,Email,Phone,Location


login.check_password
````````````````````
A Boolean (True/False) option to determine if passwords should be (required
and) checked in order to log in to MAH.

**This should always be set to True for production systems.**.

If set to False, the following log entry will be generated on start up:
::

    WARNING: Password checks are configured NOT to be enforced.

login.use_radius
````````````````
Should the login system authenticate users using Radius. If set to true,
the further login.radius_* settings are required.

login.radius_dictionary
```````````````````````
The radius dictionary to use, the default which is installed with pyrad is
generally sufficient.
::

    radius_dictionary = /usr/share/doc/python-pyrad-1.1/example/dictionary

login.radius_server
```````````````````
The radius server host name which should be used to authenticate users.
::

    radius_server = <RADIUS_hostname>

login.radius_secret
```````````````````
The radius shared secret for this application/host to use when
communicating with the Radius server.

login.radius_nas_identifer
``````````````````````````
The host name of the Network Access Server (the server running the MAH
application).
::

    radius_nas_identifer = <NAS_identifer>

login.radius_nas_ip_address
```````````````````````````
The IP of the Network Access Server (the server running the MAH
application).
::

    radius_nas_ip_address = <NAS_address>

authentication.timeout
``````````````````````
When one user authenticates another, the length of time (in seconds) that
the shared secret be visible to both users. It is prudent to have
relatively short (minutes, as opposed to hours) authentication timeouts.
::

    timeout = 600

authentication.length
`````````````````````
The length of the authentication shared secret.
::

    length = 9

authentication.length
`````````````````````
Allow variable length of the authentication shared secret in interval
<length/length-3>.
::

    variable_length = true

report.email_from
`````````````````
The from email address to use when send suspicious reports.
::

    email_from = <mutual_email>

report.email_to
```````````````
The to address to send suspicious reports to for further investigation.
::

    email_to = <admin_email>

report.email_subject
````````````````````
The subject to use for suspicious reports. Currently, there is no support
for templates/variables, so the subject must be a static string.
::

    email_subject = MAH suspicious activity report

report.smtp_server
``````````````````
The SMTP email server to use for sending suspicious activity reports.
::

    smtp_server = <SMTP_address>

application.host and application.port
`````````````````````````````````````
The host address and port of MAH application.
::

    host = 0.0.0.0
    port = 5000

application.SESSION_KEY
```````````````````````
The key used by Flask to encrypt session information stored in the cookie.

See also: `Flask Configuration Handling <http://flask.pocoo.org/docs/config/>`_

application.SESSION_COOKIE_SECURE
`````````````````````````````````
Controls if the cookie should have the secure flag set.
::

    SESSION_COOKIE_SECURE = True

See also: `Flask Configuration Handling <http://flask.pocoo.org/docs/config/>`_

application.SESSION_COOKIE_HTTPONLY
```````````````````````````````````
Controls if the cookie should have the httponly flag set.
::

    SESSION_COOKIE_HTTPONLY = True

See also: `Flask Configuration Handling <http://flask.pocoo.org/docs/config/>`_

application.PREFERRED_URL_SCHEME
````````````````````````````````
The preferred URL generation scheme, default is http.
::

    PREFERRED_URL_SCHEME = https

See also: `Flask Configuration Handling <http://flask.pocoo.org/docs/config/>`_

application.SESSION_TIMEOUT
```````````````````````````
The period of inactivity (in seconds) after which a user's session will be
invalidated and they will be required to log in again to continue.
::

    SESSION_TIMEOUT = 600

**Note:** if a application.REFRESH is set at a rate lower than
SESSION_TIMEOUT, then this may counteract the intended function of this
variable. In such a situation the (index) page would continually refresh
preventing the TIMEOUT threshold from ever being met.

application.REFRESH
```````````````````
How often (in seconds) should the index page automatically refresh. This is
::

    REFRESH = 30

logging.file_name
`````````````````
The file name of log file (if one is configured).
::

    file_name = ./mah.log

logging.file_level
``````````````````
The level of verbosity of messages logged to file. Valid levels are:
 * NONE - disable file logging
 * DEBUG - emit debugging level information and higher
 * INFO - emit general application information and higher
 * WARN - emit application warnings and higher
 * ERROR - emit error messages and higher
 * CRITICAL - emit critical messages only

::

    file_level = DEBUG

logging.syslog_level
````````````````````
The level of verbosity of messages logged via syslog. Valid levels are:
 * NONE - disable file logging
 * DEBUG - emit debugging level information and higher
 * INFO - emit general application information and higher
 * WARN - emit application warnings and higher
 * ERROR - emit error messages and higher
 * CRITICAL - emit critical messages only

::

    syslog_level = INFO

logging.syslog_facility
```````````````````````
The syslog facility to use. For more complete information on syslog
facilities, please see the `<logging.handlers.SysLogHandler documentation
<http://docs.python.org/dev/library/logging.handlers.html#logging.handlers.SysLogHandler>`_.
::

    syslog_facility =  local1

logging.syslog_host
```````````````````
The host to send syslog messages to.
::

    syslog_host = localhost

logging.syslog_port
```````````````````
The UDP port to send syslog messages to.
::

    syslog_port = 514


mah/mah/config.py
"""
A simple, shared configuration configuration object based on the standard
python `ConfigParser <http://docs.python.org/library/configparser.html>`_.

**config**
    Global configuration object. Can be treated as a dictionary with
    case-insensitive keys, or as an object with attributes that correspond to
    configuration sections. For instance, the following are equivalent::

        config['Application']
        config['application']
        config.application
        config.Application

    The value of which can be another configuration object or a string, boolean,
    integer, or float value, or a list of strings, booleans, integers or floats.

    See the :doc:`configuration` section for details.
"""
from ConfigParser import (
    RawConfigParser as ConfigParserBase,
    NoOptionError, NoSectionError
)
from traceback import format_exc
import os, re, importlib
import mah.log
from mah.database import database as db

_valid_name = re.compile(r'^[A-Za-z][A-Za-z0-9_]*\Z')

_NO_DEFAULT = object()

class ConfigParser(ConfigParserBase):
    def getint(self, section, option, default=_NO_DEFAULT):
        try:
            return ConfigParserBase.getint(self, section, option)
        except (NoSectionError, NoOptionError):
            # Missing is OK if we have a default to fall back to
            if default is _NO_DEFAULT: raise
            return default

    def getboolean(self, section, option, default=_NO_DEFAULT):
        try:
            return ConfigParserBase.getboolean(self, section, option)
        except (NoSectionError, NoOptionError):
            if default is _NO_DEFAULT: raise
            return default

    def getfloat(self, section, option, default=_NO_DEFAULT):
        try:
            return ConfigParserBase.getfloat(self, section, option)
        except (NoSectionError, NoOptionError):
            if default is _NO_DEFAULT: raise
            return default

    def get(self, section, option, default=_NO_DEFAULT):
        try:
            return ConfigParserBase.get(self, section, option)
        except (NoSectionError, NoOptionError):
            if default is _NO_DEFAULT: raise
            return default

    def section(self, name):
        return ConfigSection(self, name)

class ConfigSection(object):
    def __init__(self, cfg, name):
        self._cfg = cfg
        self.name = name

    def _list(self, option, default, cast):
        raw = self._cfg.get(self.name, option, default)
        if raw is default: return raw
        return [cast(x.strip()) for x in raw.split(',')]

    def _castint(self, raw):
        return int(raw, 10)

    def _castfloat(self, raw):
        return float(raw)

    def _castbool(self, raw):
        chk = raw.lower()
        if chk in ('true', 'on', 'yes', '1'): return True
        if chk in ('false', 'off', 'no', '0'): return False
        raise ValueError('Not a boolean: {raw}'.format(raw=raw))

    def _caststr(self, raw):
        return raw

    def int(self, option, default=_NO_DEFAULT):
        return self._cfg.getint(self.name, option, default)

    def intlist(self, option, default=_NO_DEFAULT):
        return self._list(option, default, self._castint)

    def float(self, option, default=_NO_DEFAULT):
        return self._cfg.getfloat(self.name, option, default)

    def floatlist(self, option, default=_NO_DEFAULT):
        return self._list(option, default, self._castfloat)

    def bool(self, option, default=_NO_DEFAULT):
        return self._cfg.getboolean(self.name, option, default)

    def boollist(self, option, default=_NO_DEFAULT):
        return self._list(option, default, self._castbool)

    def str(self, option, default=_NO_DEFAULT):
        return self._cfg.get(self.name, option, default)

    def strlist(self, option, default=_NO_DEFAULT):
        return self._list(option, default, self._caststr)

class AttribDict(dict):
    def __getattr__(self, key):
        lkey = key.lower()
        if lkey in self.keys():
            return self[lkey]
        raise AttributeError(
            "{cls!r} object has no attribute {attr!r}".format(
                cls=self.__class__,
                attr=key
            )
        )

    def __setattr__(self, key, value):
        if key[0] == '_':
            super(AttribDict, self).__setattr__(key, value)
        else:
            self.__setitem__(key, value)

    def __setitem__(self, key, value):
        super(AttribDict, self).__setitem__(key.lower(), value)

    def __getitem__(self, key):
        return super(AttribDict, self).__getitem__(key.lower())

    def get(self, key, default=None):
        return super(AttribDict, self).get(key.lower(), default)

    def has_key(self, key):
        return super(AttribDict, self).has_key(key.lower())

def load_module(config, src, parent, cfgparent, clsname, ifaces):
    if not _valid_name.match(config.type):
        raise ValueError(
            'Config {parent}.type must contain only alphanumerics and '
            'underscores, and must start with an alphabetic character'.format(
                parent=cfgparent
            )
        )
    try:
        config.module = importlib.import_module(
            'mah.{parent}.{pkg}'.format(parent=parent, pkg=config.type)
        )
    except Exception:
        raise RuntimeError(
            'Failed to import mah.{parent}.{pkg} - {trace}'.format(
                parent=parent,
                pkg=config.type,
                trace=format_exc()
            )
        )
    cls = getattr(config.module, clsname)
    try:
        cls.init(config, src)
    except Exception:
        raise RuntimeError(
            'Failed to initialise mah.{parent}.'
            '{pkg}.{cls} - {trace}'.format(
                parent=parent,
                pkg=config.type,
                cls=clsname,
                trace=format_exc()
            )
    )
    for iface in ifaces:
        if not callable(getattr(cls, iface, None)):
            raise RuntimeError(
                'mah.{parent}.{pkg}.{cls} is invalid - missing '
                'required interface method {method}'.format(
                    parent=parent,
                    pkg=config.type,
                    cls=clsname,
                    method=iface
                )
            )

class Config(AttribDict):
    def __init__(self, src=None):
        if src is None: return
        # We're the main config. Set ourselves up
        self.ok = False

        # Logging section
        # Do this section first and set up logging as fast as we can - even
        # before mah.config itself is set up. To do this, we pass the logging
        # section of the config to mah.log.init() so it doesn't have to wait.
        # We'll even delay raising an error until the very last second in case
        # at least one logging method is valid.

        # Note that if the configuration is unparseable, or if the logging
        # section has errors, logs might be unavailable or not where you're
        # expecting them.
        section = Config()
        rsection = src.section('logging')
        section.file_level = rsection.str('file_level', 'DEBUG')
        errors = []
        if section.file_level != 'NONE':
            if section.file_level in mah.log.LEVELS:
                section.file_name = rsection.str('file_name')
                section.file_format = rsection.str(
                    'file_format',
                    '%(asctime)s %(name)s(%(threadName)s) '
                    '%(levelname)s: %(msg)s'
                )
            else:
                errors.append('logging.file_level')
        section.syslog_level = rsection.str('syslog_level', 'DEBUG')
        if section.syslog_level != 'NONE':
            if section.syslog_level in mah.log.LEVELS:
                section.syslog_host = rsection.str('syslog_host', 'localhost')
                section.syslog_port = rsection.int('syslog_port', 514)
                section.syslog_facility = rsection.str(
                    'syslog_facility', 'user'
                )
                section.syslog_format = rsection.str(
                    'syslog_format', '%(name)s: %(msg)s'
                )
            else:
                errors.append('logging.syslog_level')
        self.logging = section
        if len(errors) < 2:
            mah.log.init(self.logging)
        if len(errors):
            raise ValueError(
                'Config {config} is invalid'
                ' - must be one of {options}'.format(
                    config=', '.join(errors),
                    options='NONE, ' + ', '.join(mah.log.LEVELS)
                ))

        # Application section
        section = Config()
        rsection = src.section('application')
        section.host = rsection.str('host')
        section.port = rsection.int('port')
        section.session_key = rsection.str('session_key')
        section.session_timeout = rsection.int('session_timeout', 600)
        section.session_cookie_secure = rsection.bool(
            'session_cookie_secure', True
        )
        section.session_cookie_httponly = rsection.bool(
            'session_cookie_httponly', True
        )
        section.preferred_url_scheme = rsection.str(
            'preferred_url_scheme', 'https'
        )
        section.refresh = rsection.int('refresh', 30)

        if section.session_timeout < 30:
            raise ValueError(
                'Config application.session_timeout must be at least 30'
            )
        if section.refresh < 10:
            raise ValueError('Config application.refresh must be at least 10')
        self.application = section

        # Authentication section
        section = Config()
        rsection = src.section('authentication')
        section.timeout = rsection.int('timeout', 300)
        section.length = rsection.int('length', 8)
        section.variable_length = rsection.bool('variable_length', True)
        if section.length < 5 or section.length > 128:
            raise ValueError(
                'Config authentication.length must be '
                'between 5 and 128 inclusive'
            )
        self.authentication = section

        # Database section
        section = Config()
        rsection = src.section('database')
        section.connect = rsection.str('connect')
        section.logsql = rsection.bool('logsql', False)
        self.database = section

        # Login section
        section = Config()
        rsection = src.section('login')
        section.type = rsection.str('type', 'none')
        load_module(
            section, rsection, 'authentication', 'login', 'Authentication',
            ('authenticate', 'for_production')
        )
        auth = section.module.Authentication
        if (not callable(getattr(auth, 'authenticate', None)) or
            not callable(getattr(auth, 'for_production', None))):
            raise RuntimeError(
                'mah.authentication.{pkg} is invalid - '
                'missing required interface'.format(
                    pkg=section.type
                )
            )
        self.login = section

        # Directory section
        section = Config()
        rsection = src.section('directory')
        section.attributes = src.strlist('attributes')
        if len(section.attributes) < 2:
            raise ValueError(
                'Config directory.attributes should '
                'be a list of at least two entries'
            )
        section.attribute_names = src.strlist('attribute_names')
        if len(section.attribute_names) < 2:
            raise ValueError(
                'Config directory.attribute_names should '
                'be a list of at least two entries'
            )
        if len(section.attribute_names) != len(section.attributes):
            raise ValueError(
                'Config directory.attribute_names should be of '
                'the same length as directory.attributes'
            )
        section.id_attribute = src.str('id_attribute')
        if section.id_attribute != section.attributes[0]:
            raise ValueError(
                'Config directory.id_attribute should '
                'match first entry of directory.attributes'
            )
        section.name_attribute = src.str('name_attribute')
        if section.name_attribute != section.attributes[1]:
            raise ValueError(
                'Config directory.name_attribute should '
                'match second entry of directory.attributes'
            )
        section.type = rsection.str('type')
        load_module(
            section, rsection, 'directory', 'directory', 'Directory',
            ('search', 'user')
        )
        self.directory = section

        # Report section
        section = Config()
        rsection = src.section('report')
        section.email_from = rsection.str('email_from', 'mah@corp.com')
        section.email_to = rsection.strlist('email_to', ['mah@corp.com'])
        section.subject = rsection.str(
            'subject', 'MAH suspicious activity report'
        )
        section.smtp_server = rsection.str('smtp_server', 'localhost')
        self.report = section
        self.ok = True

#: The config object. This is imported directly and used in various places in
#: MAH.
config = Config()
config.ok = False
config.error = 'Config not loaded.'

def load():
    global config
    src = os.environ.get('MAHCONFIG', 'mah.conf')
    raw = ConfigParser()
    try:
        with open(src, 'r') as cfg:
            raw.readfp(cfg)
    except Exception as e:
        config.error = 'Failed to open or read configuration file'
        config.exception = e
        config.trace = format_exc()
        return
    try:
        config = Config(raw)
    except Exception as e:
        config.error = 'Configuration error detected'
        config.exception = e
        config.trace = format_exc()
    try:
        db.init()
    except Exception as e:
        config.ok = False
        config.error = (
            'Database initialisation failed - configuration incorrect?'
        )
        config.exception = e
        config.trace = format_exc()
load()    
del load, load_module
